### PR TITLE
Change vm.stats.remove.batch.size to delete.batch.query.size & allow delete of volume_stats in batches

### PR DESCRIPTION
**Pull request by vishesh92**
### Description

This PR changes `vm.stats.remove.batch.size` configuration parameter to `delete.batch.query.size` to make it more generic.

Changes also allow deletion of `volume_stats` in batches using the `delete.batch.query.size` global setting.

For context, please check the conversation in PR: https://github.com/apache/cloudstack/pull/8740

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by vishesh92** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 1]
```diff
@@ -506,6 +506,10 @@ public class ConfigurationManagerImpl extends ManagerBase implements Configurati
     public static final ConfigKey<Boolean> ALLOW_DOMAIN_ADMINS_TO_CREATE_TAGGED_OFFERINGS = new ConfigKey<>(Boolean.class, "allow.domain.admins.to.create.tagged.offerings", "Advanced",
             "false", "Allow domain admins to create offerings with tags.", true, ConfigKey.Scope.Account, null);
 
+    public static final ConfigKey<Long> DELETE_BATCH_QUERY_SIZE = new ConfigKey<>("Advanced", Long.class, "delete.batch.query.size", "0",
+            "Indicates the limit applied while deleting entries in bulk. With this, ACS will run the delete query, applying the limit, as many times as necessary" +
+            " to delete all the entries. Zero (0) means that no limit will be applied, therefore, the query will run once and without limit, keeping the default behavior.", true);
+
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:512]: @vishesh92 Add upgrade path to change config name` vm.stats.remove.batch.size` => `delete.batch.query.size`

Comment (vishesh92) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:512]: This is within the same minor version. Do we need an upgrade path?

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:512]: ok, then not needed @vishesh92

**Review by rohityadavcloud — APPROVED** [review]

**Comment by vishesh92** [comment]
@blueorangutan package

**Review by sureshanaparti — APPROVED** [review]

**Comment by JoaoJandre** [comment]
Hey @vishesh92, 
As I said on #8740, I do not think that it makes sense to have a configuration with a generic name that only applies to a single scenario. This misleads the users into thinking that ACS will apply it everywhere, even if you add a note at the end of the description saying otherwise. My point on #8740 was that if we are applying the same concept on other deletes, then we can create a generic configuration. Just renaming a specific configuration to be generic and changing nothing else does seem to make much sense.

**Comment by vishesh92** [comment]
> Hey @vishesh92, As I said on #8740, I do not think that it makes sense to have a configuration with a generic name that only applies to a single scenario. This misleads the users into thinking that ACS will apply it everywhere, even if you add a note at the end of the description saying otherwise. My point on #8740 was that if we are applying the same concept on other deletes, then we can create a generic configuration. Just renaming a specific configuration to be generic and changing nothing else does not seem to make much sense.

@JoaoJandre I think I misunderstood what you meant. I thought you didn't want to increase the scope of your PR. As I mentioned earlier in PR https://github.com/apache/cloudstack/pull/8740, I know it doesn't make sense now. But I am assuming we will be doing this for other tables as well in the future. Do you have any plan on how to deprecate the global settings in the future when we have similar global setting for different tables?

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
Hi @JoaoJandre Can you review/check if your comments are addressed. It seems the config is applied for vm and volume stats now, this can be extended to other stats/tables later if required without any new config.

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by JoaoJandre** [comment]
> > Hey @vishesh92, As I said on #8740, I do not think that it makes sense to have a configuration with a generic name that only applies to a single scenario. This misleads the users into thinking that ACS will apply it everywhere, even if you add a note at the end of the description saying otherwise. My point on #8740 was that if we are applying the same concept on other deletes, then we can create a generic configuration. Just renaming a specific configuration to be generic and changing nothing else does not seem to make much sense.
> 
> @JoaoJandre I think I misunderstood what you meant. I thought you didn't want to increase the scope of your PR. As I mentioned earlier in PR #8740, I know it doesn't make sense now. But I am assuming we will be doing this for other tables as well in the future. Do you have any plan on how to deprecate the global settings in the future when we have similar global setting for different tables?

When I said I did not want to increase the scope of the PR, I was talking about applying the configuration to other deletes. 

Regarding global setting deprecation, this would be better discussed on #8970, but it could be a something like: introduce the new configuration while deprecating the older one on one release, then remove the older one on the next release.

In any case, I see you added support for the `volume_stats` table. As the description of the PR says that the idea is to make the configuration generic, will you be adding support for other tables as well?

**Comment by vishesh92** [comment]
> In any case, I see you added support for the `volume_stats` table. As the description of the PR says that the idea is to make the configuration generic, will you be adding support for other tables as well?

I am not planning to add support for other tables in this PR. I will try to add support for other tables in the future.

**Comment by sureshanaparti** [comment]
@JoaoJandre we can keep the config that is updated in this PR, and later its description would to be updated if added support for other tables. Agree?

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -506,6 +506,11 @@ public class ConfigurationManagerImpl extends ManagerBase implements Configurati
     public static final ConfigKey<Boolean> ALLOW_DOMAIN_ADMINS_TO_CREATE_TAGGED_OFFERINGS = new ConfigKey<>(Boolean.class, "allow.domain.admins.to.create.tagged.offerings", "Advanced",
             "false", "Allow domain admins to create offerings with tags.", true, ConfigKey.Scope.Account, null);
 
+    public static final ConfigKey<Long> DELETE_BATCH_QUERY_SIZE = new ConfigKey<>("Advanced", Long.class, "delete.batch.query.size", "0",
+            "Indicates the limit applied while deleting entries in bulk. With this, ACS will run the delete query, applying the limit, as many times as necessary" +
+                    " to delete all the entries. This is advised when retaining several days of records, which can lead to slowness. Zero (0) means that no limit will be applied, " +
+                    "therefore, the query will run once and without limit, keeping the default behavior. For now, this is used for deletion of vm & volume stats only.", true);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:512]:
```suggestion
                    "therefore, the query will run once and without limit, this is default value which keeps the current behavior. For now, this is used for deletion of vm & volume stats only.", true);
```

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 3]
```diff
@@ -506,6 +506,11 @@ public class ConfigurationManagerImpl extends ManagerBase implements Configurati
     public static final ConfigKey<Boolean> ALLOW_DOMAIN_ADMINS_TO_CREATE_TAGGED_OFFERINGS = new ConfigKey<>(Boolean.class, "allow.domain.admins.to.create.tagged.offerings", "Advanced",
             "false", "Allow domain admins to create offerings with tags.", true, ConfigKey.Scope.Account, null);
 
+    public static final ConfigKey<Long> DELETE_BATCH_QUERY_SIZE = new ConfigKey<>("Advanced", Long.class, "delete.batch.query.size", "0",
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:509]: @vishesh92 need this validate this config for negative values?

Comment (vishesh92) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:509]: Updated the configuration description and verified that negative values are handled in the same way as 0 is handled.

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:509]: ok, thanks.

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 3 above)
Comment (JoaoJandre) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:509]:
```suggestion
    public static final ConfigKey<Long> DELETE_QUERY_BATCH_SIZE = new ConfigKey<>("Advanced", Long.class, "delete.query.batch.size", "0",
```
I think this name makes more sense.

**Review by JoaoJandre — APPROVED** [review]
LGTM, but I think we should change the name to `delete.query.batch.size` instead of `delete.batch.query.size`. I did not test it.

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan LLtest

**Comment by sureshanaparti** [comment]
> [SF] Trillian test result (tid-10664) Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7 Total time taken: 35009 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9283-t10664-kvm-centos7.zip Smoke tests completed. 97 look OK, 34 have errors, 0 did not run

these failures are due to templates listing issue & builtin template deleted during the tests.

**Comment by sureshanaparti** [comment]
Manually verified the config change.

**Pull request merged**